File: Graph/843/n_queen.cpp
#include <iostream>
#include <cstring>
using namespace std;

const int N = 10;
int n;
bool col[N], diag[2 * N], udiag[2 * N];
char g[N][N];

void dfs(int r);

int main() {
    memset(g, '.', sizeof(g));
    scanf("%d", &n);
    dfs(0);
    return 0;
}

void dfs(int r) {
    if (r == n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) printf("%c", g[i][j]);
            printf("\n");
        }
        printf("\n");
        return;
    }
    for (int c = 0; c < n; c++) {
        if (!col[c] && !diag[r + c] && !udiag[n - r + c - 1]) {
            g[r][c] = 'Q';
            col[c] = diag[r + c] = udiag[n - r + c - 1] = true;
            dfs(r + 1);
            col[c] = diag[r + c] = udiag[n - r + c - 1] = false;
            g[r][c] = '.';
        }
    }
}

File: Graph/843/n_queen.py
def dfs(r: int, n: int, g: list, col: list, diag: list, udiag: list):
    if r == n:
        for i in range(n):
            print("".join(g[i]))
        print()
        return
    for c in range(n):
        if not col[c] and not diag[r + c] and not udiag[n - r + c - 1]:
            col[c] = diag[r + c] = udiag[n - r + c - 1] = True
            g[r][c] = "Q"
            dfs(r + 1, n, g, col, diag, udiag)
            g[r][c] = "."
            col[c] = diag[r + c] = udiag[n - r + c - 1] = False

if __name__ == "__main__":
    n = int(input())
    g = [["."] * n for _ in range(n)]
    col, diag, udiag = [False] * n, [False] * (2 * n), [False] * (2 * n)
    dfs(0, n, g, col, diag, udiag)
